dimagi/commcare-hq
# Add conditional case update status to app summary

## Product Description
Adds "save only on edit" to applicable properties in app summary, on the "form summary" page.
<img width="251" alt="image" src="https://user-images.githubusercontent.com/100609770/217334244-9b142b74-3023-40af-9ba6-2d52bbd8df5e.png">

Additionally fixes an issue where subcase "name" property actions were not shown in app summary - in form summary, case summary, or excel exports.
<img width="250" src="https://user-images.githubusercontent.com/100609770/218219553-858a837d-6c1a-4f31-9757-45308ce0bb8a.png"> => <img width="250" src="https://user-images.githubusercontent.com/100609770/218219561-4e6e7392-23ae-4b8f-8410-ba70fa13517f.png">

![image](https://user-images.githubusercontent.com/100609770/218220523-caac0018-25ae-4f43-9084-c2ac97b709b2.png) => ![image](https://user-images.githubusercontent.com/100609770/218220767-2ac16eb7-4eaf-4ed2-a04d-80f505c43f1b.png)

## Technical Summary
[USH-1693](https://dimagi-dev.atlassian.net/browse/USH-1693)
A form's update actions include an `update_mode`, either "always" or "edit", that indicate a property's conditional update status. This update adds a step in metadata that passes through the `update_mode` of an action to each of a question's `save_properties` as they are built. Changes the format of metadata question save_properties from `list [property_]` to `dict {property_: update_mode}`.

In the app summary UI, the conditional update status is shown on the form summary page when "Cases" is selected:
<img width="95" alt="image" src="https://user-images.githubusercontent.com/100609770/217337276-9c4aa178-0384-4458-b544-19950c7231f9.png">
To match the user-facing setting for this option (a single-select checkbox for "Save Only If Edited"), this info displays as a text note if the `update_mode` is "edit" - that is, if "save only if edited" is selected - otherwise, no note is shown.

This does not add conditional case update status info to the downloadable form summary given that `save_properties` are output as a simple list in the format "case type - property".


## Safety Assurance

### Safety story
This change adds an `update_mode` to a `LoadSaveProperty` when it is used in a list of save_properties. Through most of the building for this property it is passed in as a keyword argument defaulting to `None`, to ensure compatibility with other references to these same functions. 

The change in format for `add_save()` is intentionally structured so that `property in save_properties` logic will still work, with updates made where needed to accept the new format.

### Automated test coverage
Added asserts to a test under `app_manager.app_schemas.tests.test_form_metadata` to see that `update_mode` gets attached to `save_property` objects correctly. Modified one additional test to check for save property keys in the new format.

### QA Plan
[QA-4812](https://dimagi-dev.atlassian.net/browse/QA-4812)
Will send through QA to confirm the "save only if edited" notation displays on properties in app summary when it should (and only when it should), on all applicable question types, and does not cause any other unforeseen issues with the app summary.

### Rollback instructions


### Labels & Review


[USH-1693]: https://dimagi-dev.atlassian.net/browse/USH-1693?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[QA-4812]: https://dimagi-dev.atlassian.net/browse/QA-4812?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
